### ISSUE TITLE
Slower in WebWorkers than Main Thread

### ISSUE THREAD

**Issue by jvilk** [plasma-umass/doppio#401]
DoppioJVM runs slower in a WebWorker than the main browser thread in Google Chrome. I have yet to run tests in other browsers.

In Google Chrome, **45%** of DoppioJVM's runtime is attributed to `(program)`, which coalesces all of the tasks that Google Chrome does itself. These `(program)` blocks appear *between* the DoppioJVM JavaScript events. These periods are not nearly as big in the main browser thread.

[The culprit appears to be this bug, which makes our `setImmediate` polyfill significantly slower.](https://code.google.com/p/chromium/issues/detail?id=344814&can=5&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Cr%20Status%20Owner%20Summary%20OS%20Modified). DoppioJVM schedules many tasks to complete asynchronously to keep the JVM responsive to user input, and needs a fast method for scheduling work to complete as a future browser event. This bug fundamentally limits the throughput and responsiveness of DoppioJVM. The method we use in the main thread, `window.postMessage`, is unavailable in the same form in a WebWorker, so we are forced to use message ports.

I haven't checked to see if the same regression occurs in other browsers. I'll check this out when I have time.

**Comment by bertung** [comment]
Is setTimeout too slow?

**Comment by jvilk** [comment]
setTimeout  has a minimum value of 4ms, as set by the standard, so it is too slow.

**Comment by bertung** [comment]
Hmm interesting, there is no main thread in the web worker thus there is not need to delay the setTimeout to a minimum. setTimeout(method,0) must be near immediate in a web worker. Have you tried this out against the message port work around?

**Comment by jvilk** [comment]
I agree that it makes no sense for workers, but I have doubts that they would deviate from the standard. When I get a chance to run further tests, I'll post the results here.

**Comment by norzak** [comment]
Has this bug been fixed on the new release?

**Comment by jvilk** [comment]
@norzak this is not a bug in Doppio. This is a bug in Chrome. Please [star this issue](https://bugs.chromium.org/p/chromium/issues/detail?id=344814&can=5&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Cr%20Status%20Owner%20Summary%20OS%20Modified) so the Chrome developers know that this should be fixed.

**Comment by norzak** [comment]
The chrome bug looks like fixed. Any improvement on your end? does it work with doppio as expected?
